src/main/java/com/microsoft/sqlserver/jdbc/NTLMAuthentication.java

```diff
@@ -0,0 +1,906 @@
+        return 0;
+    }
+
+    /**
+     * Parses the Type 2 NTLM Challenge message from server
+     *
+     * Section 2.2.1.2 CHALLENGE_MESSAGE
+     *
+     * @param inToken
+     *        SSPI input blob
+     * @throws SQLServerException
+     */
+    private void parseNtlmChallenge(final byte[] inToken) throws SQLServerException {
+
+        context.token = ByteBuffer.wrap(inToken).order(ByteOrder.LITTLE_ENDIAN);
+
+        // verify signature
+        byte[] signature = new byte[NTLM_HEADER_SIGNATURE.length];
+        context.token.get(signature);
+        if (!Arrays.equals(signature, NTLM_HEADER_SIGNATURE)) {
+            MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_ntlmSignatureError"));
+            Object[] msgArgs = {new String(signature)};
+            throw new SQLServerException(form.format(msgArgs), null);
+        }
+
+        // verify message type
+        int messageType = context.token.getInt();
+        if (messageType != NTLM_MESSAGE_TYPE_CHALLENGE) {
+            MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_ntlmMessageTypeError"));
+            Object[] msgArgs = {messageType};
+            throw new SQLServerException(form.format(msgArgs), null);
+        }
+
+        // target name fields
+        int targetNameLen = context.token.getShort();
+        context.token.getShort(); // targetNameMaxLen
+        context.token.getInt(); // targetNameOffset
+
+        // negotiate flags
+        context.token.getInt();
+
+        // get server challenge
+        context.token.get(context.serverChallenge);
+
+        // 8 bytes reserved 32-40: all 0's
+        context.token.getLong();
+
+        // target info fields
+        int targetInfoLen = context.token.getShort();
+        context.token.getShort(); // targetInfoMaxLen
+        context.token.getInt(); // targetInfoOffset
+
+        context.token.getLong(); // version - not used
+
+        // get requested target name
+        byte[] targetName = new byte[targetNameLen];
+        context.token.get(targetName);
+
+        // verify targetInfo - was requested so should always be sent
+        context.targetInfo = new byte[targetInfoLen];
+        context.token.get(context.targetInfo);
+        if (0 == context.targetInfo.length) {
+            throw new SQLServerException(SQLServerException.getErrString("R_ntlmNoTargetInfo"), null);
+        }
+
+        // verify target info av pairs
+        ByteBuffer targetInfoBuf = ByteBuffer.wrap(context.targetInfo).order(ByteOrder.LITTLE_ENDIAN);
+        boolean done = false;
+        for (int i = 0; i < context.targetInfo.length && !done;) {
+            int id = targetInfoBuf.getShort();
+            byte[] value = new byte[targetInfoBuf.getShort()];
+            targetInfoBuf.get(value);
+
+            switch (id) {
+                case NTLM_AVID_MSVAVDNSDOMAINNAME:
+                    if (null != context.domainName && logger.isLoggable(Level.WARNING)) {

```

Comment:
Please move `logger.isLoggable(Level.WARNING)` to the next `if` statement. 
`if (!Arrays.equals(context.domainBytes, new String(value).toUpperCase().getBytes()) && logger.isLoggable(Level.WARNING))`

Reply:
actually it should be 1st